pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/HistogramAggregationFunction.java

```diff
@@ -0,0 +1,358 @@
+  private static final int INVALID_BIN = -1;
+  double[] _bucketEdges;
+  boolean _isEqualLength = false;
+  double _lower;
+  double _upper;
+  double _binLength;
+
+  public HistogramAggregationFunction(List<ExpressionContext> arguments) {
+    super(arguments.get(0));
+    int numArguments = arguments.size();
+    Preconditions.checkArgument(numArguments == 4 || numArguments == 2, "Histogram expects 2 or 4 arguments, got: %s;"
+        + "usage example: `Histogram(columnName, ARRAY[0,1,10,100])` to specify bins (0,1], (1,10], (10,1000] or "
+        + "`Histogram(columnName, 0, 1000, 10)` to specify 10 equal-length bins "
+        + "(0,100], (100,200], ..., (900,1000]", numArguments);
+    if (numArguments == 2) {
+      ExpressionContext arrayExpression = arguments.get(1);
+      Preconditions.checkArgument(
+          (arrayExpression.getType() == ExpressionContext.Type.FUNCTION) && (arrayExpression.getFunction()
+              .getFunctionName().equals(ARRAY_CONSTRUCTOR)),
+          "Please use the format of `Histogram(columnName, ARRAY[1,10,100])` to specify the bin edges");
+      _bucketEdges = parseVector(arrayExpression.getFunction().getArguments());
+      _lower = _bucketEdges[0];
+      _upper = _bucketEdges[_bucketEdges.length - 1];
+    } else {
+      _isEqualLength = true;
+      _lower = Double.parseDouble(arguments.get(1).getLiteral());
+      _upper = Double.parseDouble(arguments.get(2).getLiteral());
+      int numBins = Integer.parseInt(arguments.get(3).getLiteral());
+      Preconditions.checkArgument(_upper > _lower,
+          "The right most edge must be greater than left most edge, given %s and %s", _lower, _upper);
+      Preconditions.checkArgument(numBins > 0, "The number of bins must be greater than zero, given %s", numBins);
+      _bucketEdges = new double[numBins + 1];
+      _bucketEdges[0] = _lower;
+      _bucketEdges[numBins] = _upper;
+      _binLength = (_upper - _lower) / numBins;
+      for (int i = 1; i < numBins; i++) {
+        _bucketEdges[i] = i * _binLength + _lower;
+      }
+    }
+  }
+
+  int getNumBins() {
+    return _bucketEdges.length - 1;
+  }
+
+  int getNumEdges() {
+    return _bucketEdges.length;
+  }
+
+  private double[] parseVector(List<ExpressionContext> arrayStr) {
+    int len = arrayStr.size();
+    Preconditions.checkArgument(len > 1, "The number of bin edges must be greater than 1");
+    double[] ret = new double[len];
+    for (int i = 0; i < len; i++) {
+      ret[i] = Double.parseDouble(arrayStr.get(i).getLiteral());
+    }
+    for (int i = 1; i < len; i++) {
+      Preconditions.checkState(ret[i] > ret[i - 1], "the bin edges must be strictly increasing");
+    }
+    return ret;
+  }
+
+  /**
+   * Find the bin id for the input value. Use division for equal-length bins, and binary search otherwise.
+   * @param val input value
+   * @return bin id
+   */
+  private int getBinId(double val) {
+    if (val > _upper || val <= _lower) {
+      return INVALID_BIN;
```

Comment:
In the function `HISTOGRAM(intColumn,ARRAY[0,1,10,100,1000,10000])` values outside of (0,10000] are being dropped. Would it be possible to set up the histogram so that a bin before the first value and a bin after the last value is automatically added to catch `INVALID_BIN` values? After that if the user strictly wanted to avoid adding these two extra bins, they could add a predicate like: `SELECT HISTOGRAM(intColumn,ARRAY[0,1,10,100,1000,10000]) where intColumn > 0 and intColumn <= 10000`, so this will allow creating histogram with or without the extra bins. Same thing with `HISTOGRAM(intColumn,0,1000,10)`.

Reply:
Good ask! I was also thinking about outlier collection during the design. For `HISTOGRAM(intColumn,ARRAY[·])` pattern, users can explicitly use `HISTOGRAM(intColumn,ARRAY["-Infinity",0,1,10,100,1000,10000, "+Infinity"])`  to count the outliers on the left/right side. For the `HISTOGRAM(intColumn,0,1000,10)` I'm a bit hesitate to add this invalid bin as it might cause a semantical confusion.  

I'm trying to replicate the behavior of https://numpy.org/doc/stable/reference/generated/numpy.histogram.html

Reply:
`HISTOGRAM(intColumn,ARRAY["-Infinity",0,1,10,100,1000,10000, "+Infinity"])` looks good. For the second function  one option may be to add an optional argument at the end to indicate whether outlier bucket should be added or not:
`HISTOGRAM(intColumn,0,1000,10, true)`: add outlier bucket.
`HISTOGRAM(intColumn,0,1000,10, false)`: don't add outlier bucket.
`HISTOGRAM(intColumn,0,1000,10)` : default - don't add outlier bucket.
but the current approach is ok too.

Reply:
Yeah I have put this in the follow up and we can probably implement it together with MV. For now the user can also use `COUNT(*) where intColumn < 0 and intColumn > 10000` to count outliers.